tcmur_cmd_handler.c

```diff
@@ -423,9 +424,10 @@ static int handle_unmap_internal(struct tcmu_device *dev, struct tcmulib_cmd *or
 		state->error = true;
 
 	refcount = state->refcount;
+
 	pthread_mutex_unlock(&state->lock);
 
-	if (refcount)
+	if (refcount > 0)
```

Comment:
You do not need this do you? If refcount has gone negative then we have a bigger problem.

Reply:
Ignore that comment. I see the bug. The refcount code is a little crazy. I am back from PTO Monday, so let me try to send a patch that describes the direction for the fix and you can take it from there. It is a little hard to just describe it because as you saw it is messy in there.

Reply:
> Hey @bgly,
> 
> Attached is a completely untested patch. The idea is that unmap_state_put should always do the right thing when refcount=0. The submitting code paths then do not have to do those refcount checks. They just return TCMU_STS_ASYNC_HANDLED and unmap_state_put will complete the command when all refs are released.
> 
> [runner-fix-unmap-ref.txt](https://github.com/open-iscsi/tcmu-runner/files/2849418/runner-fix-unmap-ref.txt)

I will test now, ill let you know soon.